native/cocos/bindings/manual/jsb_conversions.h

```diff
@@ -733,8 +734,13 @@ sevalue_to_native(const se::Value &from, T **to, se::Object * /*ctx*/) { // NOLI
     } else if (data->isTypedArray()) {
         data->getTypedArrayData(&tmp, nullptr);
     } else {
-        CC_ABORT(); // bad type
-        return false;
+        auto *dataHolder = static_cast<cc::JSBNativeDataHolder *>(data->getPrivateData());
```

Comment:
Why assume that the private data is always `JSBNativeDataHolder` type?
It's better to add the JSB type by 
```cpp
void *privateData = data->getPrivateData();
if (privateData != nullptr && data->_getClass() != nullptr
  && data->_getClass() == __jsb_cc_JSBNativeDataHolder_class) {
    auto *dataHolder = static_cast<cc::JSBNativeDataHolder *>(privateData);
    tmp = dataHolder->getData();
} else {
    CC_ABORT(); // bad type
    return false;
}
```

Reply:
you better！！！

Reply:
but  **jsb_cocos_auto.h** include **jsb_conversions.h**  , circular reference!

Reply:
jsb_conversions.h
```c++

extern bool conver_native_data(se::Object *data, uint8_t **native_data);

template <typename T>
typename std::enable_if_t<!std::is_pointer<T>::value && std::is_arithmetic<T>::value, bool>
sevalue_to_native(const se::Value &from, T **to, se::Object * /*ctx*/) { // NOLINT(readability-identifier-naming)
    se::Object *data = from.toObject();
    uint8_t *tmp = nullptr;
    if (conver_native_data(data, &tmp) == false) {
        CC_ABORT(); // bad type
        return false;
    }
    *to = reinterpret_cast<T *>(tmp);
    return true;
}
```

add jsb_conversions.cpp

```c++
#include "bindings/manual/jsb_conversions.h"
#include "cocos/bindings/auto/jsb_cocos_auto.h"
#include "core/data/JSBNativeDataHolder.h"

bool conver_native_data(se::Object *data, uint8_t **native_data)
{
    bool result = false;
    if (data->isArrayBuffer()) {
        data->getArrayBufferData(native_data, nullptr);
        result = true;
    } else if (data->isTypedArray()) {
        data->getTypedArrayData(native_data, nullptr);
        result = true;
    } else {
        void *privateData = data->getPrivateData();
        if (privateData != nullptr && data->_getClass() == __jsb_cc_JSBNativeDataHolder_class) {
            auto* dataHolder = static_cast<cc::JSBNativeDataHolder*>(privateData);
            if (dataHolder != nullptr) {
                *native_data = dataHolder->getData();
                result = true;
            }
        }
    }
    return result;
}
```

It works, are you sure modify it this?

Reply:
T may be other primitive types like `uint32_t*`, `uint16_t*`, etc. 
We could use `extern se::Class * __jsb_cc_JSBNativeDataHolder_class ;` in jsb_conversion.h to resolve the circular reference.

Reply:
This commit is probably the smallest change, I'm not sure ```jsb_cocos_auto.h``` automatically generated！
move ```extern se::Class * __jsb_cc_JSBNativeDataHolder_class ;``` in ```jsb_conversion.h``` affect

Reply:
I don't mean that `move` the declaration from jsb_cocos_auto.h to `jsb_conversion.h`. 
I mean just add `extern se::Class * __jsb_cc_JSBNativeDataHolder_class;` in `jsb_conversion.h`. 
Could you have a try?

Reply:
thank you!

duplicate extern __jsb_cc_JSBNativeDataHolder_class to resolve the circular reference ```jsb_cocos_auto.h```

test It works!

Reply:
Good to know that it works. 👍